### PR TITLE
New Lint: `result_filter_map` / Mirror of `option_filter_map`

### PR DESCRIPTION
**Pull request by PartiallyTyped**
Added the `Result` mirror of `option_filter_map`.

changelog: New Lint: [`result_filter_map`]

I had to move around some code because the function def was too long 🙃.

I have also added some pattern checks on `option_filter_map`

### CODE REVIEW

**Comment by PartiallyTyped** [comment]
This partially fixes #11843, maybe @flip1995 would be interested in reviewing this?

**Comment by flip1995** [comment]
As I suggested this, I'd like to get a second opinion from a different reviewer on this :) thanks for the ping though!

**Comment by PartiallyTyped** [comment]
@flip1995 is this sufficiently close to what you had in mind for #11843? For this I just mirrored the `Option::is_some` case, but I have the hunch that's not exactly what you had in mind. Is this correct?

**Comment by flip1995** [comment]
Yes, that's what I had in mind.

**Comment by PartiallyTyped** [comment]
Sorry for the delay, life's gotten pretty busy lately.

**Comment by PartiallyTyped** [comment]
@Alexendoo is everything okay with this?

File: clippy_lints/src/methods/mod.rs [hunk 1]
```diff
@@ -3752,6 +3753,29 @@ declare_clippy_lint! {
     "using `Option.map_or(Err(_), Ok)`, which is more succinctly expressed as `Option.ok_or(_)`"
 }
 
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for iterators of `Result`s using ``.filter(Result::is_ok).map(Result::unwrap)` that may
+    /// be replaced with a `.flatten()` call.
+    ///
+    /// ### Why is this bad?
+    /// `Result` implements `IntoIterator<Item = T>`. This means that `Result` can be collected
```
Comment (Alexendoo) [inline clippy_lints/src/methods/mod.rs:3762]:
```suggestion
    /// `Result` implements `IntoIterator<Item = T>`. This means that `Result` can be flattened
```

**Comment by Alexendoo** [comment]
Just that and I think it's good to go

File: tests/ui/option_filter_map.rs [hunk 2]
```diff
@@ -3,21 +3,29 @@
...
         .into_iter()
         .map(odds_out)
         .filter(Option::is_some)
+        //~^error:`filter` for `Some` followed by `unwrap`
         .map(Option::unwrap);
     let _ = vec![1]
         .into_iter()
         .map(odds_out)
         .filter(|o| o.is_some())
+        //~^error:`filter` for `Some` followed by `unwrap`
```
[the remark this answers is not in the document]
Comment (PartiallyTyped) [reply tests/ui/option_filter_map.rs:28]: Well, this is embarrassing 😅

File: clippy_lints/src/methods/mod.rs (hunk 1 above)
Comment (Alexendoo) [reply clippy_lints/src/methods/mod.rs:3762]: The latest push seems to have lost this change

Comment (PartiallyTyped) [reply clippy_lints/src/methods/mod.rs:3762]: Must have been lost on rebasing 🙈🤦🏼‍♀️

**Comment by Alexendoo** [comment]
Thanks!

@bors r+

**Pull request merged**